human_aware_rl/imitation/behavior_cloning_tf2.py

```diff
@@ -0,0 +1,358 @@
+        else:
+            assert 'model_dir' in config, "must specify model directory if model not specified"
+            # Construct a private graph unique to this instance to run forward pass (so it doesn't conflict with rllib graphs)
+            self._graph = tf.Graph()
+            with self._graph.as_default():
+                model, bc_params = load_bc_model(config['model_dir'])
+
+        # This is here to make the class compatible with both tuples or gym.Space objs for the spaces
+        # Note: action_space = (len(Action.ALL_ACTIONS,)) is technically NOT the action space shape, which would be () since actions are scalars
+        self.observation_shape = observation_space if type(observation_space) == tuple else self.observation_space.shape
+        self.action_shape = action_space if type(action_space) == tuple else (self.action_space.n,)
+
+        # Basic check to make sure model dimensions match
+        assert self.observation_shape == bc_params['observation_shape']
+        assert self.action_shape == bc_params['action_shape']
+
+        self.model = model
+        self.stochastic = config['stochastic']
+
+    @classmethod
+    def from_model_dir(cls, model_dir, stochastic=True):
+        model, bc_params = load_bc_model(model_dir)
+        config = {
+            "bc_model" : model,
+            "bc_params" : bc_params,
+            "stochastic" : stochastic
+        }
+        return cls(bc_params['observation_shape'], bc_params['action_shape'], config)
+
+    @classmethod
+    def from_model(cls, model, bc_params, stochastic=True):
+        config = {
+            "bc_model" : model,
+            "bc_params" : bc_params,
+            "stochastic" : stochastic
+        }
+        return cls(bc_params["observation_shape"], bc_params["action_shape"], config)
+
+    def compute_actions(self, obs_batch, 
+                        state_batches=None, 
+                        prev_action_batch=None,
+                        prev_reward_batch=None,
+                        info_batch=None,
+                        episodes=None,
+                        **kwargs):
+        """
+        Computes sampled actions for each of the corresponding OvercookedEnv states in obs_batch
+
+        Args:
+            obs_batch (np.array): batch of pre-process (lossless state encoded) observations
+
+        Returns:
+            actions (list|np.array): batch of output actions shape [BATCH_SIZE, ACTION_SHAPE]
+            state_outs (list): only necessary for rnn hidden states
+            infos (dict): dictionary of extra feature batches { "action_dist_inputs" : [BATCH_SIZE, ...] }
+        """
+        # Cast to np.array if list (no-op if already np.array)        
+        obs_batch = np.array(obs_batch)
+
+        # Run the model
+        if self._graph:
+            with self._graph.as_default():
+                action_logits = self.model.predict(obs_batch)
+        else:
+            action_logits = self.model.predict(obs_batch)
```

Comment:
I had to do this because rllib wraps all  policy execution calls in their own complicated context manager, which breaks any forward pass of a tensorflow model since it can't find the tensors it needs in the graph. What I did here was create a local private graph context manager that I load my model into, and then the `with self._graph.as_default()` overrides the current graph in use for the forward pass of the model, so it doens't interfere with rllib at all. I could probably make this cleaner by creating a `context` variable that is a null context manager when graph is `None`, and thus avoid the if statement here, but I'm not sure if that would really make it any less confusing.

Reply:
Yeah, you mentioned in the call too that this had been annoying to set up because of how rllib handles things. I think in that case, for now lets leave it this way, and potentially modify it in the future if any good solution for structuring this comes to mind.

Reply:
Ended up implementing the context manager idea suggested above, which dynamically sets the graph and session based on whether tensorflow is executing eagerly or not. Basically, this makes the `BehaviorCloningPolicy` model's behavior independent of the tensorflow execution context, so that rllib can't break it when it does all its fancy under the hood optimizations. 

Also it unified the cases like you suggested and hopefully abstracts away some of the messy tensorflow details